module: Set dynamic import callback

jkrems opened the pull request:
This is an initial implementation to support dynamic import in
both scripts and modules. It's off by default since support for
dynamic import is still flagged in V8. Without setting the V8 flag,
this code won't be executed.

Because of missing support in the V8 APIs, we can't support
importing into the proper context yet. Without further changes this
might allow code to break out of the context it's running in.

/cc @bmeck 

#### TODO

- [x] Add test case to ensure import from "random contexts" is ignored
- [x] ~~Clean error handling when calling `import()` before the callback is set up~~ (can't think of a scenario where this might happen)
- [x] Cleanup code for handling import from CommonJS scripts
- [x] ~~Potentially add explicit error handling for failing import handler call~~ (just preventing a crash for now)
- [x] Ensure behavior matches browsers where appropriate (e.g. eval & friends)

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

module

jkrems commented:
This is using the 6.1 V8 API for the module callback which will change slightly in 6.2. But I expect the implementation to be fairly easy to adjust once we get there.

jkrems commented on .eslintignore:
```diff
@@ -3,6 +3,7 @@ lib/punycode.js
 test/addons/??_*/
 test/fixtures
 test/tmp*/
+test/es-module/test-esm-dynamic-import.js
```
Afaict our ESLint doesn't support dynamic import (and won't while it's still in stage 3).

jkrems commented on lib/internal/loader/Loader.js:
```diff
@@ -70,6 +71,15 @@ class Loader {
+
+  static registerImportDynamicallyCallback(loader) {
+    setImportModuleDynamicallyCallback(async (referrer, specifier) => {
+      const normalizedBase = /^file:\/\//.test(referrer) ?
```
This hack allows both CommonJS and ESM to use dynamic import and it's "guessing" which one it is. There's definitely cleaner ways to achieve this and this might break horribly in cases where the referrer is neither (e.g. `vm.runIn*Context`).

TimothyGu replied:
I would add a `path.isAbsolute()` check to the non-URL code path just to make sure it is what we expect, and throw an error otherwise. `vm.runInContext()` + friends have a way to specify file path as well through the [`filename` option](https://nodejs.org/api/vm.html#vm_vm_runincontext_code_contextifiedsandbox_options).

jkrems commented on src/module_wrap.cc:
```diff
@@ -502,6 +502,44 @@ void ModuleWrap::Resolve(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(context);
+  HandleScope handle_scope(iso);
+
+  Local<Function> import_callback =
```
There's currently an unhandled error case when dynamic import is used in an environment that doesn't have the callback defined (yet).

TimothyGu commented on test/es-module/test-esm-dynamic-import.js:
```diff
@@ -0,0 +1,11 @@
+    // Can't deepStrictEqual because ns isn't a normal object
+    assert.deepEqual(ns, { default: true });
+  }))
+  .then(null, err => { process.nextTick(() => { throw err; }); });
```
You could use `common.crashOnUnhandledRejection()` instead.

TimothyGu commented on test/es-module/test-esm-dynamic-import.js:
```diff
@@ -0,0 +1,11 @@
+// Flags: --experimental-modules --harmony-dynamic-import
```
`'use strict'`?

TimothyGu commented on src/module_wrap.cc:
```diff
@@ -502,6 +502,44 @@ void ModuleWrap::Resolve(const FunctionCallbackInfo<Value>& args) {
+    const FunctionCallbackInfo<Value>& args) {
+  Isolate* iso = args.GetIsolate();
+  Environment* env = Environment::GetCurrent(args);
+  EscapableHandleScope handle_scope(iso);
```
Does the handle scope needs to be escapable?

jkrems replied:
Don't think so, bad copy&paste. Thanks!

bmeck commented on src/module_wrap.cc:
```diff
@@ -502,6 +502,44 @@ void ModuleWrap::Resolve(const FunctionCallbackInfo<Value>& args) {
+                                                   Local<String> specifier) {
+  Isolate* iso = context->GetIsolate();
+  // TODO(jkrems): ensure proper context is used
+  Environment* env = Environment::GetCurrent(context);
```
if we could perform the check that only the main context (`env.context_`) gets `import()` handling setup, that would be good. We don't have a public exposed way in `vm` currently to setup things for creating new contexts.

jkrems replied:
I made it an explicit error condition when someone tries to use dynamic import from another context. Also added some test cases for it.

jkrems commented on test/es-module/test-esm-dynamic-import.js:
```diff
@@ -0,0 +1,66 @@
+  }));
+
+// Succeeds because it's got an valid base url
+Promise.resolve(vm.runInThisContext(`import("${relativePath}")`, {
```
Note: This will introduce a de-facto global access to `require` that doesn't exist right now. E.g. random code running in `vm.runInThisContext` will gain access to all node APIs.

Before:

```bash
% ./node <(echo "console.log(require('vm').runInThisContext('require(\"fs\")'));")
evalmachine.<anonymous>:1
require("fs")
^

ReferenceError: require is not defined
# ...
```

After:

```bash
% ./node --experimental-modules --harmony <(echo "require('vm').runInThisContext('import(\"fs\")').then(console.log);")
(node:29002) ExperimentalWarning: The ESM module loader is experimental.
{ default:
   { constants:
# ...
```

jkrems commented on lib/internal/loader/resolveRequestUrl.js:
```diff
@@ -80,6 +80,10 @@ const resolveRequestUrl = (baseURLOrString, specifier) => {
   const baseURL = normalizeBaseURL(baseURLOrString);
   let url = search(specifier, baseURL);
 
+  if (url.protocol === 'node:' && NativeModule.nonInternalExists(url.pathname)) {
```
@bmeck This was needed to support `import('fs')`. The problem was that the URL this function produces (`node:fs`) wasn't accepted as an input, leading to fragility ("better not try to normalize URLs twice!"). With this change, normalizing an already normalized URL *should* be safe.

bmeck replied:
can you explain this a bit more, right now `import("node:fs")` should throw an error from user code.

jkrems replied:
Gotcha. The general problem is that I resolve the url before calling `loader.import` and as a side effect `node:fs` makes it into first level of importing. I think I can rewrite that to never pass through `loader.import` or, alternatively, make `loader.import` take a different referrer/base url. Which would you prefer?

guybedford commented on lib/internal/loader/Loader.js:
```diff
@@ -70,6 +79,14 @@ class Loader {
+
+  static registerImportDynamicallyCallback(loader) {
+    setImportModuleDynamicallyCallback(async (referrer, specifier) => {
+      const normalizedBase = normalizeReferrerURL(referrer);
```
I take it this `referrer` path context is provided directly from V8 for scripts? What does it become for `eval` and `vm.runInThisContext` cases?

jkrems replied:
Yep, it's 1:1 the `ScriptOrigin#url` as provided by V8. For `vm.runInThisContext` it's "evalmachine.anonymous" by default but can be overridden using the `filename` option.

For both direct and indirect eval `referrer` is null - nice find! Will add test cases and make sure they fail cleanly.

jkrems replied:
Updated with test cases that document the behavior for `eval`: https://github.com/nodejs/node/pull/15713/commits/3eaf5fe46e666eed6febba7a69884d5a7ec1a7fa

guybedford replied:
Is this the final stance v8 is taking or will this API be changing at all do you know? If it is pretty much set, then I guess we should just adjust the resolver to handle undefined `referrer` as defaulting to the standard base referrer.

bmeck replied:
@guybedford this is a bug, V8 6.1 implements `import()` but 6.2 is the recommended implementations. See https://tc39.github.io/ecma262/#sec-getactivescriptormodule , which will be given instead of a string. So the C++ API changes a bit as well.

guybedford replied:
@bmeck that seems to be more sensible, thanks for the link.

guybedford commented on lib/internal/loader/Loader.js:
```diff
@@ -25,6 +27,13 @@ function getBase() {
+  if (path.isAbsolute(referrer)) {
+    return getURLFromFilePath(referrer);
+  }
+  return referrer;
```
Is this check needed because the `referrer` is a file path for scripts and a URL for modules?

jkrems replied:
Yep, exactly.

jkrems commented:
Alright, cleaned up the tests to not repeat the checks every single time and introduced `loadNamespaceForRequest` which allows `import("node:fs")` to fail and `import("fs")` to succeed while still sharing most of the code with import statements.

Fishrock123 commented:
assigning in a futile attempt to remember this

bmeck commented:
+1 with the knowledge that we will have to update when V8 6.2 lands.

jkrems commented on src/module_wrap.cc:
```diff
@@ -517,7 +517,7 @@ static MaybeLocal<Promise> ImportModuleDynamically(Local<Context> context,
         OneByteString(iso, "invalid_context"));
       if (resolver->Reject(context, error).IsJust()) {
-        return resolver.As<Promise>();
+        return handle_scope.Escape(resolver.As<Promise>());
```
@TimothyGu I admit that I'm not 100% familiar with proper `HandleScope` handling - but I assumed that I had to go back to an `EscapableHandleScope` because I'm now creating a new object (`Promise::Resolver::New` above) in this code path? Happy to revert if this isn't necessary. :)

TimothyGu replied:
@jkrems Yep I believe this is necessary now. Thanks for checking!

jkrems commented:
Added one last small commit to clean up `MaybeLocal` and scope escaping from my understanding of those. Removed WIP because this should be good to land now (with @bmeck's disclaimer that the handling of referrer and C++ signatures will change once 6.2 lands).

jkrems commented:
@MylesBorins I'm hesitant to land this in case 6.2 is "just around the corner". Do we have a good feeling yet if 6.2 will land before end of October?

targos commented:
According to the [Chromium release schedule](https://www.chromium.org/developers/calendar), V8 6.2 will become stable on Oct 17th. We should be landing it shortly after.

jkrems commented:
Alright, will wait for 6.2 and adjust before merge. Shouldn't affect most of the implementation and might prevent annoying failures/conflicts.

jdalton commented:
Is there a reference for the `referrer` behavior?

jdalton commented on test/es-module/test-esm-dynamic-import.js:
```diff
@@ -0,0 +1,96 @@
+expectInvalidUrlError(vm.runInThisContext(`import("${relativePath}")`));
+
+// Rejects because it's got an invalid referrer URL
+expectInvalidUrlError(vm.runInThisContext(`import("${targetURL}")`));
```
Is this rejecting an absolute file url?
```js
expectInvalidUrlError(vm.runInThisContext(`import("${targetURL}")`))
```

TimothyGu replied:
@jdalton It seems like it.

jdalton replied:
That seems really odd. Absolute paths don't have ambiguity like relative ones so it seems weird to error in this scenario (when they normally don't error).

jkrems replied:
`vm` is somewhat of a special case because it doesn't really have an equivalent in browsers. But since I can't find a single case in my browsers testing where even relative imports don't work, I guess it would make sense to align `vm.runInThisContext` with `eval` / `new Function` / etc.?

Originally I thought that allowing `import("file:///absolute.mjs")` would poke a hole into `vm.runIn{This,New}Context` that doesn't exist today so I was hesitant to enable it. It's less of an issue in `vm.runInThisContext` where `process.mainModule.require` is accessible but it did seem like a bad precedent to expose it.

bmeck replied:
[`vm` already lets you specify the source origin](https://github.com/nodejs/node/blob/d597317a209fcf9e0c233093b2adc8f71129b9ef/src/node_contextify.cc#L663). Is it too much to ask people to set that?

jdalton replied:
>  Is it too much to ask people to set that?

I think `vm.runInThisContext('import(someAbsolutePath)')` should "just work". No extra options/config needed.

bmeck replied:
I'm not convinced it should just work if you don't specify where the `import` is coming from. What do loader hooks see for example? I can't put `null` since that would be the same the module/script referrer for multiple module records.

bmeck replied:
I need to think on if people need to include extra metadata for their loading like CORS via https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-crossorigin which also affects anything in this vein of duplicate referrers.

jdalton replied:
> I'm not convinced it should just work if you don't specify where the import is coming from. What do loader hooks see for example?

Have it get context similar to how `require` does. *Something to keep in mind when making these choices is every time we make ESM more cumbersome its another reason to stay on CJS.

bmeck replied:
@jdalton `require` is not available by default in a `vm.Script`. You would need to pass one in.

jdalton replied:
> `require` is not available by default in a `vm.Script`. You would need to pass one in.

I get it. I [showed an example](https://github.com/nodejs/node/pull/15713#discussion_r149586177) earlier of simply assigning `require` to a local var. The gist is the same though.

jkrems replied:
@bmeck For `runInThisContext`, you *can* pretty reliably get a `require` if you know the absolute path of the thing you're trying to import (`process.mainModule.require("<absolute path")`). A local var won't help but anything global (like `process.mainModule` or @jdalton's example above) works.

TimothyGu commented:
@jdalton Do you mean reference to a specification? The HTML Standard only talks about browsers, which always has a current URL to use as referrer, so unfortunately the empty referrer behavior does not have a reference. Otherwise, the *[resolve a module specifier](https://html.spec.whatwg.org/multipage/webappapis.html#resolve-a-module-specifier)* algorithm may be what you are looking for, called from [HostImportModuleDynamically](https://html.spec.whatwg.org/multipage/webappapis.html#hostimportmoduledynamically(referencingscriptormodule,-specifier,-promisecapability)).

jdalton commented on test/es-module/test-esm-dynamic-import.js:
```diff
@@ -0,0 +1,96 @@
+}));
+
+// Rejects because it's got an invalid referrer URL
+expectInvalidUrlError(vm.runInThisContext(`import("${relativePath}")`));
```
It looks like `require` handle cases like this without a base url (uses the cwd()?)
Would that be an option here?
```js
expectInvalidUrlError(vm.runInThisContext(`import("${relativePath}")`));
```

TimothyGu replied:
@jdalton 

> It looks like `require` handle cases like this without a base url

Can you provide a test case where `require` works?

jdalton replied:
```shell
$ touch foo.js
$ node
vm.runInThisContext(`require('./foo.js')`)
```

TimothyGu replied:
@jdalton Ah, that `require` is actually created by the REPL as a global function with the CWD as base. This function is not available in CJS modules executed by `node file.js`.

jdalton replied:
Out of the repl `touch bar.js` then
```js
//bar.js
vm = require("vm")
req = require
console.log(vm.runInThisContext(`req('./foo.js')`))
```
Then `node bar.js`.

jdalton replied:
Node's `require` works because of `path.resolve` use which will use `cwd()`.

TimothyGu replied:
@jdalton That has nothing to do with the CWD, but rather with where `bar.js` is... which in your case is the CWD.

jdalton replied:
@TimothyGu Cool! It is of its context though and does work. Would dynamic import be able to cue off of `import.meta.url` to get its bearings similar to `require()`?

bmeck replied:
`import.meta.url` is a writable property in [web spec](https://html.spec.whatwg.org/#hostgetimportmetaproperties), you should use a private bound variable instead if you are an environment.

jdalton replied:
> `import.meta.url` is a writable property in web spec, you should use a private bound variable instead if you are an environment.

Yep yep!

jkrems replied:
@jdalton Just to make sure I understand the `bar.js` example correctly: You are assigning the `require` from the module closure to a global `req` and then use that to require something. I'm not sure I follow how this relates to `import`. `import` cannot be assigned or forwarded. The only context we have is the eval/script context of the code running `import`. So I don't think it would make sense to make relative imports work where the importing script pretty explicitly doesn't have a location.

jdalton replied:
@jkrems IMO if it's doable in cjs it would be rad to be doable in esm. I'm for consistency though, so if that means deprecating some aspect in cjs that's better than leaving something doable in cjs but more cumbersome in esm.

jkrems replied:
You can do the same thing you showed above in esm:

```js
//bar.js
vm = require("vm")
req = specifier => import(specifier);
vm.runInThisContext(`req('./foo.js')`).then(console.log)
```

The above should just work.

jdalton replied:
@jkrems Nice! On a separate note, any thoughts on import() with absolute file urls erroring? That still seems like an odd one to error on.

jkrems replied:
Well, it *would* be possible to make it do something but it would mean introducing a completely new mode ("resolve but without an actual base") and I traced it through a handful of layers where things might need to change. And it's a feature that requires disabling all checks along the way for something that *normally* shouldn't happen.

The shorter version: I agree that it's something that's reasonable to expect and I'd be happy to put it on the list. My proposal right now would be "let it throw in the initial PR, carefully make it work in a follow-up".

jdalton replied:
@jkrems 
>The shorter version: I agree that it's something that's reasonable to expect and I'd be happy to put it on the list. My proposal right now would be "let it throw in the initial PR, carefully make it work in a follow-up".

I can dig that!

jkrems commented:
After testing in latest Chrome (Canary) & Safari (Tech Preview) - all the eval-ish cases, including relative imports, are expected to work:

```
import { runCase } from './nested/run-case.js';

runCase('relative', () => import('./a.js'));
runCase('relative, direct eval', () => eval("import('./a.js')"));
runCase('relative, indirect eval', () => eval.call(null, "import('./a.js')"));
runCase('relative, new Function', () => new Function("return import('./a.js')")());
runCase('absolute', () => import('http://localhost:4000/examples/a.js'));
runCase('absolute, direct eval', () => eval("import('http://localhost:4000/examples/a.js')"));
runCase('absolute, indirect eval', () => eval.call(null, "import('http://localhost:4000/examples/a.js')"));
runCase('absolute, new Function', () => new Function("return import('http://localhost:4000/examples/a.js')")());
```

Just noting it down here for future reference. Still waiting for 6.2.436+ to make it into any node branch so we get https://github.com/v8/v8/commit/dbfe4a49d88f6655ed31285a40b63786ab1e8e49 and can clean up this change.

targos commented:
> Still waiting for 6.2.436+ to make it into any node branch so we get v8/v8@dbfe4a4 and can clean up this change

Then you're waiting for 6.3 because the 6.2 branch was cut at an earlier minor version.
If that commit alone is enough, an option is to cherry-pick it.

jkrems commented:
@targos Ah! Thanks. I was afraid that might be the case. Would cherry-pick happen on the V8 side or would we apply it ourselves? In either case I could look into it.

addaleax commented:
@jkrems I don’t think this could be cherry-picked on V8’s side since it is not a bugfix, but you could apply it here if that’s all that’s needed.

jkrems commented:
Gave it a shot here: https://github.com/nodejs/node/pull/16889

jkrems commented on test/es-module/test-esm-dynamic-import.js:
```diff
@@ -0,0 +1,96 @@
+expectInvalidUrlError(vm.runInThisContext(`import("${targetURL}")`));
+
+// Rejects because it's running in the wrong context
+expectInvalidContextError(
```
Double-checked in Chrome and at least for `<iframe>`, each context gets it's own module map. Which I guess would be possible in the future for node as well, but for now throwing seems fair - at least as the default behavior.

jkrems commented:
Mostly thanks to the newer V8, all the eval test cases now match browser behavior. Reorganized the test file to have some meaningful groupings. This is now using the official V8 API that shipped in Chrome 64 (e.g. `Local<ScriptOrModule>`).

MylesBorins commented:
hey @jkrems 

Would you be able to give a high level update on where this PR is currently at

Which version of V8 do we need on master for this to work? Is this supported in both ESM + common? When do you think we will be able to land this?

jkrems commented:
> Which version of V8 do we need on master for this to work?

This requires the new dynamic import API which is included in 6.3 but can be added to 6.2 by cherry-picking https://github.com/v8/v8/commit/dbfe4a49d88f6655ed31285a40b63786ab1e8e49. This is why this PR works against master even though we don't have 6.3 yet (see https://github.com/nodejs/node/pull/16889).

> Is this supported in both ESM + common?

Yes, this implementation works in both ESM + common. For the latter we currently treat a CJS script at `/foo/bar.js` as if it would have the URL `file:///foo/bar.js`. That conversion is [implemented in `normalizeReferrerURL `](https://github.com/nodejs/node/pull/15713/files#diff-e6f2798544dae1f6ec863621e745489aR31).

> When do you think we will be able to land this?

Since the required V8 patch is now on master, this *should* be good to land once it got some approvals. The reviews that are on the PR right now refer to an earlier state where it was using an older version of the V8 API. No *major* changes happened since but enough that I would want some explicit 👍s.

jkrems commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11734/

jkrems commented:
Huh, looking into why [it fails in `test/linux-debug` only](https://ci.nodejs.org/job/node-test-commit-linux-linked/nodes=ubuntu1604_sharedlibs_debug_x64/401/console):

```
19:09:27 not ok 1988 es-module/test-esm-dynamic-import
19:09:27   ---
19:09:27   duration_ms: 1.312
19:09:27   severity: crashed
19:09:27   stack: |-
19:09:27     oh no!
19:09:27     exit code: CRASHED (Signal: 4)
19:09:27   ...
```

TimothyGu commented on src/module_wrap.cc:
```diff
@@ -550,6 +550,62 @@ void ModuleWrap::Resolve(const FunctionCallbackInfo<Value>& args) {
+    if (maybe_resolver.ToLocal(&resolver)) {
+      // TODO(jkrems): Turn into proper error object w/ code
+      Local<Value> error = v8::Exception::Error(
+        OneByteString(iso, "invalid_context"));
```
Are there plans to support dynamic `import()` in VM contexts? If so, I'd rather the error be something to the sense of "Dynamic import() callback not registered for current context" than "invalid\_context".

jkrems replied:
It... depends. Especially in [a future where Realms are a thing](https://github.com/tc39/proposal-realms), we'd want some sort of support for having per-context module maps. So this error is meant to be temporary, pending a decision around how and where to handle module imports in contexts other than the main one.

So depending on where that discussion goes, the actual error wording (and code) might be different. So I would rather defer introducing an opinion here for now. :)

TimothyGu replied:
Thanks for the link to the proposal. I see reasons for your ambivalence about being too opinionated. However, I'd still prefer something a bit more user-friendly than "invalid\_context", maybe just "TypeError: import() called in VM context"? We can worry about `err.code` later.

jkrems commented:
* Added a missing `.Escape` that messed up the heap.
* Added a clearer error message when using `import()` from code outside of the main context.

jkrems commented:
It looks like that Windows failure is a general trend in recent builds (not just of this PR):

```
(node:2340) ExperimentalWarning: The http2 module is an experimental API.
assert.js:42
  throw new errors.AssertionError({
  ^

AssertionError [ERR_ASSERTION]: <Buffer ff d8 ff e0 00 10 4a 46 49 46 00 01 01 01 00 48 00 48 00 00 ff ed 4a 9c 50 68 6f 74 6f 73 68 6f 70 20 33 2e 30 00 38 42  deepStrictEqual <Buffer ff d8 ff e0 00 10 4a 46 49 46 00 01 01 01 00 48 00 48 00 00 ff ed 4a 9c 50 68 6f 74 6f 73 68 6f 70 20 33 2e 30 00 38 42 
    at ServerHttp2Stream.stream.on.common.mustCall (c:\workspace\node-test-binary-windows\test\parallel\test-http2-client-upload.js:28:14)
    at ServerHttp2Stream.<anonymous> (c:\workspace\node-test-binary-windows\test\common\index.js:522:15)
    at ServerHttp2Stream.emit (events.js:164:20)
    at endReadableNT (_stream_readable.js:1062:12)
    at _combinedTickCallback (internal/process/next_tick.js:137:11)
    at process._tickCallback (internal/process/next_tick.js:179:9)
```

jkrems commented:
ping @nodejs/http2: Is there a known issue with test-http2-client-upload failing on Windows w/ VS2015? Looking at the last 10+ failed builds (https://ci.nodejs.org/job/node-test-commit-windows-fanned/), they mostly seem to fail with the same reason (found at least a handful).

jasnell commented:
Possibly. Will investigate

jkrems commented:
Alright, best theory I have is that I got stuck with a bad base commit from master. After rebase I get a passing build. Gonna land this later today.

TimothyGu commented on src/env.h:
```diff
@@ -332,6 +332,7 @@ class ModuleWrap;
   V(vm_parsing_context_symbol, v8::Symbol)                                    \
   V(url_constructor_function, v8::Function)                                   \
   V(write_wrap_constructor_function, v8::Function)                            \
+  V(host_import_module_dynamically_callback, v8::Function)                    \
```
Nit when landing: alphabetize

jkrems commented:
Landed in https://github.com/nodejs/node/commit/59e48329d00cd91f6836cd91bcb8aca92acac1f6

MylesBorins commented:
While this patch lands cleanly on v9.x, it relies on https://github.com/nodejs/node/commit/3a4fe7791e1c2a1bcb7ac62ffbbcdefe41a45b92 which has a change to the API signature of a V8 api, potentially resulting in ABI breakages.

We will need a backport put together with a non abi breaking version of 3a4fe7791e

/cc @nodejs/v8

jkrems commented:
Trying to come up with a way to make it ABI compatible but I assume `V8_INLINE ScriptOrigin` means that the current implementation would have been inlined into a potential `.node` binary, leaving the `host_defined_options` field uninitialized at best..?

jkrems commented:
One possible way to backport would be https://github.com/nodejs/node/pull/17823.

MylesBorins commented:
backported to v8.x in 453077fc6e...a0c7df8a10
